airbyte-cdk/python/airbyte_cdk/v2/concurrency/stream_group.py

```diff
@@ -0,0 +1,122 @@
+import asyncio
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import List, Mapping, Any, Generic, TypeVar, Iterable, AsyncIterable
+
+from airbyte_cdk.sources.utils.record_helper import stream_data_to_airbyte_message
+
+from airbyte_cdk.sources.connector_state_manager import HashableStreamDescriptor
+from airbyte_protocol.models import ConfiguredAirbyteCatalog, AirbyteMessage, Type as MessageType, AirbyteStateMessage, AirbyteStateType, \
+    AirbyteStreamState, StreamDescriptor
+
+from airbyte_cdk.v2 import Stream
+from airbyte_cdk.v2.concurrency import PartitionedStream, PartitionDescriptor, ConcurrencyPolicy, AsyncRequester
+from airbyte_cdk.v2.concurrency.concurrent_utils import consume_async_iterable
+from airbyte_cdk.v2.state_obj import State
+
+PartitionType = TypeVar('PartitionType', bound=PartitionDescriptor)
+StateType = TypeVar('StateType', bound=State)
+
+
+@dataclass
+class StreamAndPartition(Generic[PartitionType]):
+    stream: PartitionedStream
+    partition_descriptor: PartitionType
+
+
+class ConcurrentStreamGroup(ABC, Generic[PartitionType]):
+    requester: AsyncRequester
+    concurrency_policy: ConcurrencyPolicy
```

Comment:
in hindsight I think the requester should probably hold the concurrency policy because it should be the one coordinating retrieval of the partitions